File: coj/3236.py
from sys import stdin, stdout

def rdl():
	return stdin.readline().replace('\n', '')

def rdls():
	return rdl().split(' ')

def find(parents, i):
	if (parents[i] == i):
		return i
	else:
		parents[i] = find(parents,parents[i])
		return parents[i]

def union_find(parents, num_sets, u, v):
	up = find(parents, u)
	vp = find(parents, v)
	if (up != vp):
		num_sets[0] -= 1
		parents[up] = vp

def main():
	cases = int(rdl())
	while(cases):
		line = rdls()
		n, e = int(line[0]), int(line[1])
		parents = [x for x in range(n)]
		num_sets = [n]
		for i in range(e):
			line = rdls()
			u, v = int(line[0]) - 1, int(line[1]) - 1
			union_find(parents, num_sets, u, v)
		num_sets = num_sets[0]
		result = int(((num_sets-1)*(num_sets))/2)
		stdout.write('{}\n'.format(result))
		cases-=1

if __name__ == '__main__':
	main()

File: coj/3236.cpp
#include <iostream>
using namespace std;

#define MAX 10000

int parents[MAX];
int n_sets;

int _find(int i) {
	return ((parents[i] == i) ? i : parents[i] = _find(parents[i]));
}

void _union_find(int u, int v) {
	int pu = _find(u);
	int pv = _find(v);
	if (pu != pv) {
		n_sets--;
		parents[pu] = pv;
	}
}

int main() {
	int cases;
	cin>>cases;
	while(cases--) {
		int n,m;
		cin>>n>>m;
		n_sets = n;
		for (int i = 0; i < n; i++)
			parents[i] = i;
		while(m--) {
			int u, v;
			cin>>u>>v;
			_union_find(u-1, v-1);
		}
		long result = ((n_sets-1)*n_sets)/2;
		cout<<result<<endl;
	}
}